framework-plugins/lisk-framework-forger-plugin/src/forger_plugin.ts

```diff
@@ -288,97 +293,113 @@ export class ForgerPlugin extends BasePlugin {
-			isUpdated = true;
-
 			// eslint-disable-next-line @typescript-eslint/no-floating-promises
 			this._webhooks.handleEvent({
 				event: 'forger:block:created',
 				timestamp: Date.now(),
 				payload: { reward, forgerAddress, height },
 			});
+			await setForgerInfo(this._forgerPluginDB, forgerAddressBinary, { ...forgerInfo });
 		}
 
-		for (const trx of payload) {
-			if (trx.type === VoteTransaction.TYPE) {
-				for (const vote of (trx.asset as Asset).votes) {
-					const registeredDelegateIndex = this._forgersList.findIndex(
-						forger => forger.address === vote.delegateAddress,
-					);
-					if (registeredDelegateIndex !== -1) {
-						const delegateVoteIndex = forgerInfo.votesReceived.findIndex(aVote =>
-							aVote.address.equals(Buffer.from(vote.delegateAddress, 'base64')),
-						);
-						if (delegateVoteIndex === -1) {
-							forgerInfo.votesReceived.push({
-								address: Buffer.from(vote.delegateAddress, 'base64'),
-								amount: BigInt(vote.amount),
-							});
-						} else {
-							forgerInfo.votesReceived[delegateVoteIndex].amount += BigInt(vote.amount);
-						}
-						isUpdated = true;
-					}
-				}
-			}
-		}
-
+		await this._addVotesReceived(payload);
 		await this._updateMissedBlock(encodedBlock);
-
-		if (isUpdated) {
-			await setForgerInfo(this._forgerPluginDB, forgerAddressBinary, { ...forgerInfo });
-		}
 	}
 
-	private async _decrementForgerInfo(encodedBlock: string): Promise<void> {
+	private async _revertForgerInfo(encodedBlock: string): Promise<void> {
 		const {
 			forgerAddress,
 			forgerAddressBinary,
 			header: { reward },
 			payload,
 		} = this._getForgerHeaderAndPayloadInfo(encodedBlock);
 		const forgerInfo = await getForgerInfo(this._forgerPluginDB, forgerAddressBinary);
-		let isUpdated = false;
 
 		if (this._forgersList.find(forger => forger.address === forgerAddress)) {
 			forgerInfo.totalProducedBlocks -= 1;
 			forgerInfo.totalReceivedRewards -= BigInt(reward);
 			forgerInfo.totalReceivedFees -= this._getFee(payload, encodedBlock);
-			isUpdated = true;
+			await setForgerInfo(this._forgerPluginDB, forgerAddressBinary, { ...forgerInfo });
 		}
 
+		await this._revertVotesReceived(payload);
+	}
+
+	private _getForgerReceivedVotes(payload: ReadonlyArray<TransactionJSON>): ForgerReceivedVotes {
+		const forgerReceivedVotes: ForgerReceivedVotes = {};
+
 		for (const trx of payload) {
 			if (trx.type === VoteTransaction.TYPE) {
-				for (const vote of (trx.asset as Asset).votes) {
-					const delegateVoteIndex = forgerInfo.votesReceived.findIndex(aVote =>
-						aVote.address.equals(Buffer.from(vote.delegateAddress, 'base64')),
+				const senderAddress = getAddressFromPublicKey(Buffer.from(trx.senderPublicKey, 'base64'));
+				(trx.asset as Asset).votes.reduce((acc: ForgerReceivedVotes, curr) => {
+					const registeredDelegateIndex = this._forgersList.findIndex(
+						forger => forger.address === curr.delegateAddress,
 					);
```

Comment:
The `_forgersList` is loaded once on start and does not change. If we can convert it to `hash map` with address as key, we can avoid few of such loops in different places and code will look more readable.